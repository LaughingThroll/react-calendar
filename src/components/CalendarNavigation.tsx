import React, { useEffect, useCallback } from 'react'

enum ENavigationDirection {
  ARROW_LEFT = 'ArrowLeft',
  ARROW_RIGHT = 'ArrowRight',
}

enum EMonthNavigation {
  NEXT = 'next',
  PREV = 'prev',
}

type monthNavigation = EMonthNavigation.NEXT | EMonthNavigation.PREV

interface INavigation {
  date: Date
  onChangeMonth: (date: Date) => void
}

const Navigation: React.FC<INavigation> = ({ date, onChangeMonth }) => {
  const formatedDate = date.toLocaleDateString('en-US', { month: 'long', year: 'numeric' })

  const changeMonth = useCallback(
    (value: monthNavigation) => {
      let num = 0
      if (value === EMonthNavigation.NEXT) num = 1
      if (value === EMonthNavigation.PREV) num = -1
      onChangeMonth(new Date(date.setMonth(date.getMonth() + num)))
      // this.setState({ currentDate, daysInMonth: daysInMonth(currentDate) })
    },
    [date, onChangeMonth],
  )

  const switchMonthOnKey = useCallback(
    (e: KeyboardEvent) => {
      if (e.key === ENavigationDirection.ARROW_LEFT) changeMonth(EMonthNavigation.PREV)
      if (e.key === ENavigationDirection.ARROW_RIGHT) changeMonth(EMonthNavigation.NEXT)
    },
    [changeMonth],
  )

  useEffect(() => {
    window.addEventListener('keyup', switchMonthOnKey)

    return () => {
      window.removeEventListener('keyup', switchMonthOnKey)
    }
  }, [switchMonthOnKey])

  return (
    <div className="calendar-nav">
      <button
        className="calendar-nav__btn calendar-nav__btn--prev-month"
        onClick={() => changeMonth(EMonthNavigation.PREV)}
      ></button>
      <div className="calendar-nav__current-date">{formatedDate}</div>
      <button
        className="calendar-nav__btn calendar-nav__btn--next-month"
        onClick={() => changeMonth(EMonthNavigation.NEXT)}
      ></button>
    </div>
  )
}

export default React.memo(Navigation)
