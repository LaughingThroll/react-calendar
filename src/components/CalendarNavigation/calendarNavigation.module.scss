@import 'styles/utils/mixin';

.calendarNavigation {
  display: flex;
  justify-content: space-between;
  align-items: center;
  max-width: 226px;

  &__currentDate {
    @extend %font-bold;
  }

  &__btn {
    width: 13px;
    height: 13px;
    outline: none;
    background-position: center;
    background-repeat: no-repeat;
    background-color: transparent;
    transition: transform 0.1s linear;

    &_prevMonth {
      background-image: url('../../assets/images/icons/arrow--left.svg');
    }

    &_nextMonth {
      background-image: url('../../assets/images/icons/arrow--right.svg');
    }

    &:hover {
      transform: scale(1.6);
    }
  }
}
