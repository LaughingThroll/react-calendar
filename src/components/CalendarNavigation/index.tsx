import React, { useEffect, useCallback } from 'react'
import styles from './calendarNavigation.module.scss'

enum NavigationDirection {
  ARROW_LEFT = 'ArrowLeft',
  ARROW_RIGHT = 'ArrowRight',
}

enum MonthNavigation {
  NEXT = 'next',
  PREV = 'prev',
}

type monthNavigationUnion = MonthNavigation.NEXT | MonthNavigation.PREV

export interface Navigation {
  date: Date
  onChangeMonth: (date: Date) => void
  classNames?: string[]
}

const {
  calendarNavigation,
  calendarNavigation__currentDate,
  calendarNavigation__btn,
  calendarNavigation__btn_prevMonth,
  calendarNavigation__btn_nextMonth,
} = styles

const CalendarNavigation: React.FC<Navigation> = ({ date, onChangeMonth, classNames = [] }) => {
  const formatedDate = date.toLocaleDateString('en-US', { month: 'long', year: 'numeric' })

  const changeMonth = useCallback(
    (value: monthNavigationUnion) => {
      let num = 0
      if (value === MonthNavigation.NEXT) num = 1
      if (value === MonthNavigation.PREV) num = -1
      onChangeMonth(new Date(date.setMonth(date.getMonth() + num)))
    },
    [date, onChangeMonth]
  )

  const switchMonthOnKey = useCallback(
    (e: KeyboardEvent) => {
      if (e.key === NavigationDirection.ARROW_LEFT) changeMonth(MonthNavigation.PREV)
      if (e.key === NavigationDirection.ARROW_RIGHT) changeMonth(MonthNavigation.NEXT)
    },
    [changeMonth]
  )

  useEffect(() => {
    window.addEventListener('keyup', switchMonthOnKey)

    return () => {
      window.removeEventListener('keyup', switchMonthOnKey)
    }
  }, [switchMonthOnKey])

  return (
    <div className={[calendarNavigation, ...classNames].join(' ')}>
      <button
        className={[calendarNavigation__btn, calendarNavigation__btn_prevMonth].join(' ')}
        onClick={() => changeMonth(MonthNavigation.PREV)}
      ></button>
      <div className={calendarNavigation__currentDate}>{formatedDate}</div>
      <button
        className={[calendarNavigation__btn, calendarNavigation__btn_nextMonth].join(' ')}
        onClick={() => changeMonth(MonthNavigation.NEXT)}
      ></button>
    </div>
  )
}

export default React.memo(CalendarNavigation)
