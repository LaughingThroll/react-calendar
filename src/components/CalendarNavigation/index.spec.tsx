import { mount, ReactWrapper } from 'enzyme'

import CalendarNavigation, { Navigation } from './'

const setUp = (props: Navigation) => mount(<CalendarNavigation {...props} />)

describe('CalendarNavigation render', () => {
  it('default render', () => {
    const date = new Date('2021-04-21')
    const props = { date, onChangeMonth: () => {} }
    const tree = setUp(props)

    const keyMap: { [key: string]: any } = {}
    window.addEventListener = jest.fn((e, key) => (keyMap[e] = key))
    window.removeEventListener = jest.fn((e, key) => (keyMap[e] = undefined))

    expect(tree).toMatchSnapshot()
    expect(tree.text().includes('April 2021')).toBe(true)

    tree.unmount()
    expect(window.removeEventListener).toHaveBeenCalled()
  })

  it('other date render', () => {
    const date = new Date('2021-06-21')
    const props = { date, onChangeMonth: jest.fn() }
    const tree = setUp(props)

    expect(tree).toMatchSnapshot()
    expect(tree.text().includes('June 2021')).toBe(true)
  })
})

describe('CalendarNavigation click changeMonth', () => {
  let date: Date
  let props: Navigation
  let tree: ReactWrapper

  beforeEach(() => {
    date = new Date('2021-04-21')
    props = { date, onChangeMonth: jest.fn() }
    tree = setUp(props)
  })

  it('first button must trigger onChangeMonth', () => {
    const $btn = tree.find('button').first()

    $btn.simulate('click')

    expect(props.onChangeMonth).toHaveBeenCalled()
  })

  it('second button must trigger onChangeMonth', () => {
    const $btn = tree.find('button').last()

    $btn.simulate('click')

    expect(props.onChangeMonth).toHaveBeenCalled()
  })
})

describe('CalendarNavigation keyboard event', () => {
  let date: Date
  let props: Navigation
  let keyMap: { [key: string]: any }

  beforeEach(() => {
    date = new Date('2021-04-21')
    props = { date, onChangeMonth: jest.fn() }
    keyMap = {}

    window.addEventListener = jest.fn((e, key) => {
      keyMap[e] = key
    })

    setUp(props)
  })

  it('ArrowLeft must trigger onChangeMonth', () => {
    keyMap.keyup({ key: 'ArrowLeft' })

    expect(props.onChangeMonth).toHaveBeenCalled()
  })

  it('ArrowRight must trigger onChangeMonth', () => {
    keyMap.keyup({ key: 'ArrowRight' })

    expect(props.onChangeMonth).toHaveBeenCalled()
  })
})
