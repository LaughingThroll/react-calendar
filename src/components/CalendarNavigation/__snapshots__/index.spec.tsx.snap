// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarNavigation render default render 1`] = `
<Memo(CalendarNavigation)
  date={2021-04-21T00:00:00.000Z}
  onChangeMonth={[Function]}
>
  <div
    className="calendarNavigation"
  >
    <button
      className="calendarNavigation__btn calendarNavigation__btn_prevMonth"
      onClick={[Function]}
    />
    <div
      className="calendarNavigation__currentDate"
    >
      April 2021
    </div>
    <button
      className="calendarNavigation__btn calendarNavigation__btn_nextMonth"
      onClick={[Function]}
    />
  </div>
</Memo(CalendarNavigation)>
`;

exports[`CalendarNavigation render other date render 1`] = `
<Memo(CalendarNavigation)
  date={2021-06-21T00:00:00.000Z}
  onChangeMonth={[MockFunction]}
>
  <div
    className="calendarNavigation"
  >
    <button
      className="calendarNavigation__btn calendarNavigation__btn_prevMonth"
      onClick={[Function]}
    />
    <div
      className="calendarNavigation__currentDate"
    >
      June 2021
    </div>
    <button
      className="calendarNavigation__btn calendarNavigation__btn_nextMonth"
      onClick={[Function]}
    />
  </div>
</Memo(CalendarNavigation)>
`;
